content/posts/a_confusing_build_failure.md

```diff
@@ -0,0 +1,122 @@
+`final`, unless explicitly declared `open`, the error above was thrown.
+
+Although the culprit is found, we still have not figured out why the tests passed locally but failed on the CI server. 
+
+## Maybe Mockito can mock final classes after all
+
+If Mockito cannot mock final classes, then why do local tests pass? Well, sure there must be some way to mock a final class, and that's exactly what happens when the `org.mockito:mockito-inline` dependency is added to the gradle build file. 
+
+The problem is that the failing module did not compile with the dependency above. Instead, a submodule did declare mockito-inline on its `build.gradle` but not as a transitive dependency, meaning that final classes could be mocked on that submodule only, not on the top module whose tests were failing. 
+
+This could easily be fixed by either declaring mockito-inline as transitive, or adding it on the top module, but that would not explain why the test resources are compiled differently on the CI server than locally.
+
+Based on gradle files, on the CI server the build failed as expected, but locally mockito-inline was compiled on the top module (where it shouldn't) and it wrongfully passed. This, however, begs the following question: how can the CI server be compiled and tested in a different way than the local build?
+
+It turns out that different commands are employed to compile and test the source code. On the CI server the project is compiled and tested with gradle commands, thus the mockito-inline is not compiled and the test that mocks Kotlin classes fails. On the other hand the
+local builds used the test runner that is natively built into Android Studio, which happens to compile all libraries irrespectively of how they are declared on each module's `gradle.build`.
+
+## But why did mockito-inline was used on the first place?
+
+At last, the mystery has been solved but let's dig a little deeper. Why did a test try to mock an otherwise concrete class? And when/why was the mockito-inline dependency introduced first?
+
+The Kotlin class was mocked just as a filler object for the constructor. No verifications were performed on that mocked object - neither was it used to stub a value. Hence, the concrete class should be used as is or in case of an interface, which cannot be instantiated, a dummy test double. If a concrete class was mocked because it was difficult to instantiate(for example because it has too many dependencies on its constructor) that is a moment where it is needed to pause and
```

Comment:
```suggestion
The Kotlin class was mocked just as a filler object for the constructor. No verifications were performed on that mocked object - neither was it used to stub a value. Hence, the concrete class should be used as is or in case of an interface, which cannot be instantiated, a dummy test double. If, however, a concrete class was mocked because it was difficult to instantiate (e.g. it has too many dependencies on its constructor), then this is a moment where one needs to pause and
```